src/IntervalSets.jl

```diff
@@ -4,29 +4,73 @@ module IntervalSets
 
 using Base: @pure
 import Base: eltype, convert, show, in, length, isempty, isequal, issubset, ==, hash,
-             union, intersect, minimum, maximum, extrema, range, ⊇
+             union, intersect, minimum, maximum, extrema, range, ⊇, mean, median
 
 using Compat
 using Compat.Dates
 
-export AbstractInterval, ClosedInterval, ⊇, .., ±, ordered, width
+export AbstractInterval, Interval, OpenInterval, ClosedInterval,
+            ⊇, .., ±, ordered, width, leftendpoint, rightendpoint,
+            infimum, supremum
 
-abstract type AbstractInterval{T} end
+abstract type AbstractInfiniteSet{T} end
+abstract type AbstractInterval{T} <: AbstractInfiniteSet{T} end
 
-include("closed.jl")
+
+"The left endpoint of the interval."
+leftendpoint(d::AbstractInterval) = d.left
+
+"The right endpoint of the interval."
+rightendpoint(d::AbstractInterval) = d.right
+
+"A tuple containing the left and right endpoints of the interval."
+endpoints(d::AbstractInterval) = (leftendpoint(d), rightendpoint(d))
+
+"Is the interval closed at the left endpoint?"
+function isleftclosed end
+
+"Is the interval closed at the right endpoint?"
+function isrightclosed end
+
+
+
+# open_left and open_right are implemented in terms of closed_* above, so those
+# are the only ones that should be implemented for specific intervals
+"Is the interval open at the left endpoint?"
+isleftopen(d::AbstractInterval) = !isleftclosed(d)
+
+"Is the interval open at the right endpoint?"
+isrightopen(d::AbstractInterval) = !isrightclosed(d)
+
+# Only closed if closed at both endpoints, and similar for open
+isclosed(d::AbstractInterval) = isleftclosed(d) && isrightclosed(d)
+isopen(d::AbstractInterval) = isleftopen(d) && isrightopen(d)
+
+function infimum(d::AbstractInterval{T}) where T
+    a = leftendpoint(d)
+    b = rightendpoint(d)
+    a > b && throw(ArgumentError("Infimum not defined for empty intervals"))
+    a
+end
+
+function supremum(d::AbstractInterval{T}) where T
+    a = leftendpoint(d)
+    b = rightendpoint(d)
+    a > b && throw(ArgumentError("Supremum not defined for empty intervals"))
+    b
+end
+
+mean(d::AbstractInterval) = one(eltype(d))/2 * (leftendpoint(d) + rightendpoint(d))
+median(d::AbstractInterval) = mean(d)
```

Comment:
I'm not sure about this definition. Typically (not always, granted) the median of some set is an element of the set, but that isn't guaranteed when using a mean. Maybe this is fine, I dunno.